controllers/nstemplateset/nstemplateset_controller_test.go

```diff
@@ -968,6 +968,63 @@ func TestDeleteNSTemplateSet(t *testing.T) {
 		AssertThatNSTemplateSet(t, namespaceName, username, r.Client).
 			DoesNotHaveFinalizer() // finalizer was not added and nothing else was done
 	})
+
+	t.Run("NSTemplateSet not deleted until namespace is deleted", func(t *testing.T) {
+		// given an NSTemplateSet resource and 1 active user namespaces ("dev")
+		nsTmplSet := newNSTmplSet(namespaceName, username, "advanced", withNamespaces("abcde11", "dev"), withDeletionTs(), withClusterResources("abcde11"))
+		devNS := newNamespace("advanced", username, "dev", withTemplateRefUsingRevision("abcde11"))
+
+		r, fakeClient := prepareController(t, nsTmplSet, devNS)
+		req := newReconcileRequest(namespaceName, username)
+
+		// only add deletion timestamp, but not delete
+		fakeClient.MockDelete = func(ctx context.Context, obj client.Object, opts ...client.DeleteOption) error {
+			if obj, ok := obj.(*corev1.Namespace); ok {
+				deletionTs := metav1.Now()
+				obj.DeletionTimestamp = &deletionTs
+				if err:= r.Client.Update(context.TODO(), obj); err !=nil {
+					return err
+				}
+			}
+			return nil
+		}
+		// first reconcile, deletion is triggered
+		_, err := r.Reconcile(context.TODO(), req)
+		require.NoError(t, err)
+		//then second reconcile to check if namespace has actually been deleted
+		_, err = r.Reconcile(context.TODO(), req)
+		require.NoError(t, err)
+
+		// get the first namespace and check that it is not deleted
+		firstNSName := fmt.Sprintf("%s-dev", username)
+		AssertThatNamespace(t, firstNSName, r.Client).DoesExist()
```

Comment:
You are actually testing your MockDelete() here. Not the controller. Instead of checking that the namespace still exists it's better to test that the corresponding namespace has the deletion timestamp set in lines 993 and 996. So you would make sure that the delete function was called.